File: packages/sdk-python/src/openparser/_telemetry.py
"""SDK telemetry headers attached to every outbound request."""

from __future__ import annotations

import platform
import sys

SDK_LANGUAGE = "python"
SDK_VERSION = "0.0.1"


def _detect_runtime() -> str:
    impl = platform.python_implementation().lower()
    name = "python" if impl == "cpython" else impl
    v = sys.version_info
    return f"{name}-{v.major}.{v.minor}.{v.micro}"


def _detect_os() -> str:
    system = platform.system().lower() or "unknown"
    machine = platform.machine().lower() or "unknown"
    return f"{system}-{machine}"


def build_telemetry_headers() -> dict[str, str]:
    runtime = _detect_runtime()
    os_name = _detect_os()
    return {
        "X-OpenParser-Sdk": SDK_LANGUAGE,
        "X-OpenParser-Sdk-Version": SDK_VERSION,
        "X-OpenParser-Sdk-Runtime": runtime,
        "X-OpenParser-Sdk-Os": os_name,
        "User-Agent": f"openparser-sdk-python/{SDK_VERSION} ({runtime}; {os_name})",
    }

File: packages/sdk-typescript/src/telemetry.ts
/**
 * SDK telemetry headers attached to every outbound request.
 */

export const SDK_LANGUAGE = 'typescript';
export const SDK_VERSION = '0.0.4';

function detectRuntime(): string {
  const g = globalThis as unknown as {
    Bun?: { version: string };
    Deno?: { version?: { deno?: string } };
  };
  if (g.Bun?.version) return `bun-${g.Bun.version}`;
  if (g.Deno?.version?.deno) return `deno-${g.Deno.version.deno}`;
  if (typeof process !== 'undefined' && process.versions?.node) {
    return `node-${process.versions.node}`;
  }
  return 'browser';
}

function detectOs(): string {
  if (typeof process !== 'undefined' && process.platform) {
    return `${process.platform}-${process.arch}`;
  }
  return 'browser';
}

export function buildTelemetryHeaders(): Record<string, string> {
  const runtime = detectRuntime();
  const os = detectOs();
  return {
    'X-Openparser-Sdk': SDK_LANGUAGE,
    'X-Openparser-Sdk-Version': SDK_VERSION,
    'X-Openparser-Sdk-Runtime': runtime,
    'X-Openparser-Sdk-Os': os,
    'User-Agent': `openparser-sdk-typescript/${SDK_VERSION} (${runtime}; ${os})`,
  };
}
